tests/topotests/lib/bgp.py

```diff
@@ -5245,13 +5247,15 @@ def verify_fib_default_route(tgen, topo, dut, routes, expected_nexthop):
             )
             return False
 
-    if is_ipv4_default_route_found and is_ipv6_default_route_found:
-        return True
+    if expect_both:
```

Comment:
This API is to verify both ipv4 and ipv6 routes. With "expect_both" implementation, if expect_both is not passed to API then API would pass if either of IPv4 or IPv6 route is found. We should verify both IPv4 and IPv6 routes as this test is having dual config(IPv4 and IPv6).

Reply:
`expect_both` defaults to `True` so the behavior is not changed if it's not passed. This parameter is added for cases (see changes in `test_verify_default_originate_after_BGP_and_FRR_restart_p2` where `expect_both=False`) where neither IPv4 nor IPv6 route is expected. That is, `!ipv4 && !ipv6` equals to `!(ipv4 || ipv6)`, not `!(ipv4 && ipv6)`.

Reply:
You are right in test test_verify_default_originate_after_BGP_and_FRR_restart_p2:  Verification is neither of IPv4 or IPv6 route should present in fib. 

But now when expected_both=False is passed, and if (is_ipv4_default_route_found=True and is_ipv6_default_route_found=False) or (is_ipv4_default_route_found=False and is_ipv6_default_route_found=True) then also API would return True. Which will cause test failure. Now luckily we are not hitting this scenario.

Please check once.

Reply:
In this case, `expected_both=False` is used along with `expected=False`, so `verify_fib_default_route` (decorated by `@retry`) will loop until returning `False` (both default routes disappear) or timeout. I think it's probably working as expected?

Reply:
I don't quite get it... When `expected_both` is not passed, it defaults to `True`, and it works the same way as before.

Reply:
Current code implementation:
```   
    if expect_both:
        if is_ipv4_default_route_found and is_ipv6_default_route_found:
            return True
    else:
        if is_ipv4_default_route_found or is_ipv6_default_route_found:
            return True

    logger.error("Default Route for ipv4 and ipv6 address family is not found in FIB ")
    return False
```
here else condition would hit if expected_both=True(default value) but we have
or operation b/w is_ipv4_default_route_found and is_ipv6_default_route_found,
which means if either of these is True then also API would return True. But we want 
to verify both(v4 and v6) to be True as per test steps.

I did not really understand the need of expected_both parameter here, because we 
want both at anytime to be present or not.

How below code was not solving the purpose:
```
    if is_ipv4_default_route_found and is_ipv6_default_route_found:
           return True
    else:
        logger.error(
            "Default Route for ipv4 and ipv6 address family is not found in FIB "
        )
        return False
```
Please help me to understand if I am missing anything.

Reply:
> here else condition would hit if expected_both=True(default value)

Why the `else` branch would hit if `expected_both=True`?

> I did not really understand the need of expected_both parameter here, because we want both at anytime to be present or not.

See `test_verify_default_originate_after_BGP_and_FRR_restart_p2` where the logic is inverted. It's expecting `(not ipv4) and (not ipv6)`, which can be transformed to <code>not (ipv4 **or** ipv6)</code>. If theres were no `expected_both`, it would be <code>not (ipv4 **and** ipv6)</code>. Thus `expected_both=False` comes here to turn the **and** into **or**.

Reply:
> 
Yes my bad, I meant when expect_both=False, then else branch would hit.

Test test_verify_default_originate_after_BGP_and_FRR_restart_p2 where the logic is inverted. It's expecting verify_fib_default_route() API to return False. Which can not be transformed to not (ipv4 or ipv6). Because we are not doing any validation in API, it will only return True/False, validation/assert would be done in test case on the basis of API return value.

With you implementation test will pass even when any of v4 or v6 default route is present.  But it should be both v4 or v6 either present or not. Recheck once. Thank you.

Reply:
First, I want to make sure we are talking about this test case:

        step("After restarting the BGP daomon Verify the default originate ")
        DEFAULT_ROUTES = {"ipv4": "0.0.0.0/0", "ipv6": "0::0/0"}
        result = verify_fib_default_route(
            tgen,
            topo,
            dut="r2",
            routes=DEFAULT_ROUTES,
            expected_nexthop=DEFAULT_ROUTE_NXT_HOP_R3,
            expect_both=False,
            expected=False,
        )
        assert (
            result is not True
        ), "Testcase {} : Failed \n IBGP default route should be prefeered over EBGP \n Error: {}".format(
            tc_name, result
        )

If there were no `expect_both=False`, `verify_fib_default_route` would return `False` if IPv4 default route **is in** FIB while IPv6 is not, and the case will pass. This is not desired, because we want to ensure the FIB contains neither IPv4 nor IPv6 default route from R3.

Reply:
To make it more clear, see the original code:

    if is_ipv4_default_route_found and is_ipv6_default_route_found:
        return True
    else:
        logger.error(
            "Default Route for ipv4 and ipv6 address family is not found in FIB "
        )
        return False

When `is_ipv4_default_route_found=True` and `is_ipv6_default_route_found=False`, it returns `False`, causing the above test case to pass unexpectedly.

Reply:
No problem.